osu.Game/Beatmaps/BeatmapOnlineLookupQueue.cs

```diff
@@ -0,0 +1,222 @@
+
+            cacheDownloadRequest.PerformAsync();
+        }
+
+        private bool checkLocalCache(BeatmapSetInfo set, BeatmapInfo beatmap)
+        {
+            // download is in progress (or was, and failed).
+            if (cacheDownloadRequest != null)
+                return false;
+
+            // database is unavailable.
+            if (!storage.Exists(cache_database_name))
+                return false;
+
+            if (string.IsNullOrEmpty(beatmap.MD5Hash)
+                && string.IsNullOrEmpty(beatmap.Path)
+                && beatmap.OnlineBeatmapID == null)
+                return false;
+
+            try
+            {
+                using (var db = new SqliteConnection(DatabaseContextFactory.CreateDatabaseConnectionString("online.db", storage)))
+                {
+                    db.Open();
+
+                    using (var cmd = db.CreateCommand())
+                    {
+                        cmd.CommandText = "SELECT beatmapset_id, beatmap_id, approved, user_id FROM osu_beatmaps WHERE checksum = @MD5Hash OR beatmap_id = @OnlineBeatmapID OR filename = @Path";
+
+                        cmd.Parameters.Add(new SqliteParameter("@MD5Hash", beatmap.MD5Hash));
+                        cmd.Parameters.Add(new SqliteParameter("@OnlineBeatmapID", beatmap.OnlineBeatmapID ?? (object)DBNull.Value));
+                        cmd.Parameters.Add(new SqliteParameter("@Path", beatmap.Path));
+
+                        using (var reader = cmd.ExecuteReader())
+                        {
+                            if (reader.Read())
+                            {
+                                var status = (BeatmapSetOnlineStatus)reader.GetByte(2);
+
+                                beatmap.Status = status;
+                                beatmap.BeatmapSet.Status = status;
+                                beatmap.BeatmapSet.OnlineBeatmapSetID = reader.GetInt32(0);
+                                beatmap.OnlineBeatmapID = reader.GetInt32(1);
+
+                                if (beatmap.Metadata != null)
+                                    beatmap.Metadata.AuthorID = reader.GetInt32(3);
+
+                                if (beatmap.BeatmapSet.Metadata != null)
+                                    beatmap.BeatmapSet.Metadata.AuthorID = reader.GetInt32(3);
+
+                                logForModel(set, $"Cached local retrieval for {beatmap}.");
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                logForModel(set, $"Cached local retrieval for {beatmap} failed with {ex}.");
+            }
+
+            return false;
+        }
+
+        private void logForModel(BeatmapSetInfo set, string message) =>
+            ArchiveModelManager<BeatmapSetInfo, BeatmapSetFileInfo>.LogForModel(set, $"{nameof(BeatmapOnlineLookupQueue)}] {message}");
```

Comment:
nit: some imbalanced square brackets here, i think?

```suggestion
            ArchiveModelManager<BeatmapSetInfo, BeatmapSetFileInfo>.LogForModel(set, $"[{nameof(BeatmapOnlineLookupQueue)}] {message}");
```